rpc/rpc.go

```diff
@@ -1755,28 +1755,59 @@ func (c *lockedConn) sendMessage(ctx context.Context, build func(rpccp.Message)
 // 'in' channel, until the context is canceled or an error occurs. The first
 // time RecvMessage() returns an error, its results will still be sent on the
 // channel.
-func reader(ctx context.Context, in chan<- incomingMessage, t transport.Transport) {
-	for {
-		inMsg, err := t.RecvMessage()
-		incoming := incomingMessage{
-			IncomingMessage: inMsg,
-			err:             err,
+func (c *Conn) reader(ctx context.Context, in chan<- transport.IncomingMessage) func() error {
```

Comment:
I do not understand why or how this got more complicated; it seems like just wrapping the old implementation in backgroundTask should do the trick?

Reply:
Yeah, I woke up today and had the same though.  I've reverted this back to the original `incomingMessage` approach.